### PR TITLE
Raises informative warning for unicode issues during `cruft update`

### PR DESCRIPTION
**Pull request by tdhopper**
Raises informative error for issues discussed in #61 and #81

### CODE REVIEW

**Comment by HappyEinara** [comment]
Hey folks, just a heads-up: I'm currently drafting a PR for #103: a "reverse diff" which will get updates in your project vs the template, rather than updates in the template vs the project. My motivation is to help when writing a template from the ground up using a toy project to develop and fix features, and also to see if anything added in a newish real-world project would be worth backporting into the template.

The reason I mention this is that to do this at all sensibly, cruft will need to ignore `.git` and anything excluded by `.gitignore` in the project repo.

So you may find that my solution, if accepted, might overlap with this issue and possibly prove helpful to you.

**Comment by samj1912** [comment]
Thanks for the PR! Could you also add a test case for this change :)

**Comment by tdhopper** [comment]
I'm working on a test

File: tests/test_api.py [hunk 1]
```diff
@@ -71,6 +71,20 @@ def test_update_stores_checkout_value(value, tmpdir):
     assert json.load((project_dir / ".cruft.json").open("r"))["checkout"] == value
 
 
+def test_fail_on_non_unicode_in_diff(tmpdir):
+    tmpdir.chdir()
+    project_dir = cruft.create(
+        "https://github.com/tdhopper/cookiecutter-test",
```
Comment (samj1912) [inline tests/test_api.py:77]: You can use https://github.com/cruft/cookiecutter-test to create any test data btw.

**Review by samj1912 — COMMENTED** [review]
Feel free to submit a PR to https://github.com/cruft/cookiecutter-test with any test data

File: tests/test_api.py (hunk 1 above)
Comment (tdhopper) [reply tests/test_api.py:77]: @samj1912 What's the process for getting a new branch on there?

Comment (samj1912) [reply tests/test_api.py:77]: I can temporarily add you to the repository. One second.

Comment (samj1912) [reply tests/test_api.py:77]: Sent!

**Comment by tdhopper** [comment]
I realized in writing the test that the main reason we were having an issue was the post-gen hook which initializes git

**Comment by tdhopper** [comment]
Okay, I have no idea how the tests fails sometime 😅

**Comment by tdhopper** [comment]
It passed on 3.9, then I re-ran it and it failed

**Comment by tdhopper** [comment]
@samj1912 I created `test-things` and `post-gen-git-init` can be deleted in cookiecutter-test. Sorry about the mess.

**Comment by tdhopper** [comment]
Okay, I thought I fixed it with a new test dataset, but it's still nondeterministic. 

I can run this command multiple times locally 

```
rm -rf test && cruft create --directory=dir -y --checkout a0c6601138c7800ee9ec9ff005dfbfc15f2b15a4 https://github.com/cruft/cookiecutter-test && cd test && cruft update --checkout initialize-as-git-in-hook -y; cd /tmp
```

and sometimes it presents the new error, but sometimes I get this 

```
[main (root-commit) ee6380d] initial
 1 file changed, 3 insertions(+)
 create mode 100644 README.md
Checking patch .git/COMMIT_EDITMSG...
error: invalid path '.git/COMMIT_EDITMSG'

Project directory may have *.rej files reflecting merge conflicts with the update. Please resolve those conflicts manually.
Good work! Project's cruft has been updated and is as clean as possible!
```

I think something is going on with Git that is causes `.git/COMMIT_EDITMSG` to persist too long.

**Comment by samj1912** [comment]
> Okay, I thought I fixed it with a new test dataset, but it's still nondeterministic.
> 
> I can run this command multiple times locally
> 
> ```
> rm -rf test && cruft create --directory=dir -y --checkout a0c6601138c7800ee9ec9ff005dfbfc15f2b15a4 https://github.com/cruft/cookiecutter-test && cd test && cruft update --checkout initialize-as-git-in-hook -y; cd /tmp
> ```
> 
> and sometimes it presents the new error, but sometimes I get this
> 
> ```
> [main (root-commit) ee6380d] initial
>  1 file changed, 3 insertions(+)
>  create mode 100644 README.md
> Checking patch .git/COMMIT_EDITMSG...
> error: invalid path '.git/COMMIT_EDITMSG'
> 
> Project directory may have *.rej files reflecting merge conflicts with the update. Please resolve those conflicts manually.
> Good work! Project's cruft has been updated and is as clean as possible!
> ```
> 
> I think something is going on with Git that is causes `.git/COMMIT_EDITMSG` to persist too long.

would it be easier to just check in file which have content that is not utf8 decode able?

**Comment by tdhopper** [comment]
@samj1912 When I do that, I get a result like this:

```Checking patch bad-data...
error: cannot apply binary patch to 'bad-data' without full index line
error: bad-data: patch does not apply

Project directory may have *.rej files reflecting merge conflicts with the update. Please resolve those conflicts manually.
Good work! Project's cruft has been updated and is as clean as possible!```

I don't understand enough about git and binary data to understand what the stuff about `full index line` means.

Apologies again for all the branches in cruft-test....

**Comment by tdhopper** [comment]
@samj1912 I'm pretty stumped, I'd take whatever help you can offer

**Comment by samj1912** [comment]
Should be fixed now.

**Review by samj1912 — APPROVED** [review]

**Comment by tdhopper** [comment]
@samj1912 Any hope of a new release soon so we can get this change"?

**Comment by samj1912** [comment]
I will try to cut a release later today or tomorrow :)

**Comment by tdhopper** [comment]
@samj1912 Let me know if there's anything I can do to help make a release happen. (Not being passive aggressive, I'm actually happy to help 😄 )

**Comment by samj1912** [comment]
Should be released now - sorry for the delay. Was taking some time off :)

**Pull request merged**